src/main/java/org/jetbrains/research/kotlincodesmelldetector/utils/FirUtils.kt

```diff
@@ -0,0 +1,87 @@
+package org.jetbrains.research.kotlincodesmelldetector.utils
+
+import com.intellij.openapi.project.Project
+import org.jetbrains.kotlin.fir.declarations.FirFile
+import org.jetbrains.kotlin.fir.declarations.FirRegularClass
+import org.jetbrains.kotlin.fir.declarations.FirSimpleFunction
+import org.jetbrains.kotlin.fir.declarations.FirVariable
+import org.jetbrains.kotlin.fir.java.FirProjectSessionProvider
+import org.jetbrains.kotlin.fir.psi
+import org.jetbrains.kotlin.fir.resolve.dfa.FirControlFlowGraphReferenceImpl
+import org.jetbrains.kotlin.fir.resolve.dfa.cfg.*
+import org.jetbrains.kotlin.fir.resolve.transformers.FirTotalResolveTransformer
+import org.jetbrains.kotlin.idea.fir.FirModuleResolveStateImpl
+import org.jetbrains.kotlin.idea.fir.getOrBuildFir
+import org.jetbrains.kotlin.utils.DFS
+
+fun extractFirSimpleFunctions(firFile: FirFile): List<FirSimpleFunction> {
```

Comment:
Here we also could use visitor. In your code you are able to find only top-level functions and members of top-level classes